sw/device/tests/sim_dv/keymgr_key_derivation.c

```diff
@@ -0,0 +1,333 @@
+  flash_cfg_region(&info_region);
+
+  uint32_t address = FLASH_MEM_BASE_ADDR + page_id * flash_get_page_size();
+
+  ASSERT_EQZ(flash_page_erase(address, kInfoPartition));
+  ASSERT_EQZ(flash_write(address, kInfoPartition, data, size));
+
+  for (size_t i = 0; i < size; ++i) {
+    uint32_t got_data;
+    ASSERT_EQZ(flash_read(address + i * sizeof(uint32_t), kInfoPartition,
+                          /*size=*/1, &got_data));
+    CHECK(got_data == data[i]);
+  }
+  info_region.rd_en = false;
+  info_region.prog_en = false;
+  info_region.erase_en = false;
+  flash_cfg_region(&info_region);
+}
+
+static void init_flash(void) {
+  // setup default access for data partition
+  flash_default_region_access(/*rd_en=*/true, /*prog_en=*/true,
+                              /*erase_en=*/true);
+  // Initialize flash secrets.
+  write_info_page(kFlashInfoPageIdCreatorSecret, kCreatorSecret,
+                  ARRAYSIZE(kCreatorSecret));
+  write_info_page(kFlashInfoPageIdOwnerSecret, kOwnerSecret,
+                  ARRAYSIZE(kOwnerSecret));
+}
+
+static const dif_pwrmgr_wakeup_reason_t kWakeUpReasonPor = {
+    .types = 0,
+    .request_sources = 0,
+};
+
+/** Check wakeup reason to determine what to do during the test */
+static bool compare_wakeup_reasons(const dif_pwrmgr_wakeup_reason_t *lhs,
+                                   const dif_pwrmgr_wakeup_reason_t *rhs) {
+  return lhs->types == rhs->types &&
+         lhs->request_sources == rhs->request_sources;
+}
+
+static void aon_timer_wakeup_config(dif_aon_timer_t *aon_timer,
+                                    uint32_t wakeup_threshold) {
+  // Make sure that wake-up timer is stopped.
+  CHECK_DIF_OK(dif_aon_timer_wakeup_stop(aon_timer));
+
+  // Make sure the wake-up IRQ is cleared to avoid false positive.
+  CHECK_DIF_OK(dif_aon_timer_irq_acknowledge(aon_timer,
+                                             kDifAonTimerIrqWkupTimerExpired));
+
+  bool is_pending;
+  CHECK_DIF_OK(dif_aon_timer_irq_is_pending(
+      aon_timer, kDifAonTimerIrqWkupTimerExpired, &is_pending));
+  CHECK(!is_pending);
+
+  CHECK_DIF_OK(dif_aon_timer_wakeup_start(aon_timer, wakeup_threshold, 0));
+}
+
+/**
+ * Busy-wait until the DAI is done with whatever operation it is doing.
+ */
+static dif_otp_ctrl_t otp;
+
+static void wait_for_dai(void) {
+  while (true) {
+    dif_otp_ctrl_status_t status;
+    CHECK_DIF_OK(dif_otp_ctrl_get_status(&otp, &status));
+    if (bitfield_bit32_read(status.codes, kDifOtpCtrlStatusCodeDaiIdle)) {
+      return;
+    }
+    LOG_INFO("Waiting for DAI...");
+  }
+}
+
+/**
+ * Lock otp secret 2 partition so that on reboot flash seeds can be
+ * automatically loaded
+ */
+static void lock_otp_secret_partition2(void) {
+  // initialize otp
+  CHECK_DIF_OK(dif_otp_ctrl_init(
+      mmio_region_from_addr(TOP_EARLGREY_OTP_CTRL_CORE_BASE_ADDR), &otp));
+
+  CHECK_DIF_OK(dif_otp_ctrl_dai_digest(&otp, kDifOtpCtrlPartitionSecret2, 0));
+
+  wait_for_dai();
+}
```

Comment:
This could be another testutil, to which you pass `otp` context and `partition` as args.

Reply:
added a function in testutil